File: MATLAB_models/fluid_class.m
classdef fluid_class
    
    properties
        fluid
        type
        handle
        fname
        tab
        p
        T
        h
        s
        q
        cp
        rho
        mdot
        vdot % m3/s
    end
    
    methods
        function obj = fluid_class(fluid_name)
            obj.fluid = fluid_name;

            switch obj.fluid
                case 'nitrate salt'
                    obj.type = 'tab';
                    obj.fname = '.\data\SolarSalt.dat';
                case'chloride salt'
                    obj.type = 'tab';
                    obj.fname = '.\data\ChlorideSalt.dat';
                case'Therminol VP1'
                    obj.type = 'tab';
                    obj.fname = '.\data\TherminolVP1.dat';
                otherwise
                    obj.type = 'cp';
            end

            switch obj.type
                case 'cp'
                    % Check python installed
                    a = pyenv;

                    % Check CoolProp package works
                    b= py.CoolProp.CoolProp.PropsSI('T','P',101325,'Q',0,'Water');
                case 'tab'
                    obj.tab = readmatrix(obj.fname);
            end

            
        end
        
        % Calculate remaining fluid properties given two properties
        function obj = calc_fluid_props(obj,mode)
            switch mode
                case {'pT','Tp'}
                    for i = 1:numel(obj.T)
                        obj.s(i) = py.CoolProp.CoolProp.PropsSI('S','P',obj.p(i)*1e5,'T',obj.T(i)+273.15,obj.fluid)/1e3;
                        obj.h(i) = py.CoolProp.CoolProp.PropsSI('H','P',obj.p(i)*1e5,'T',obj.T(i)+273.15,obj.fluid)/1e3;
                        obj.q(i) = py.CoolProp.CoolProp.PropsSI('Q','P',obj.p(i)*1e5,'T',obj.T(i)+273.15,obj.fluid);
                        obj.rho(i) = py.CoolProp.CoolProp.PropsSI('D','P',obj.p(i)*1e5,'T',obj.T(i)+273.15,obj.fluid);
                        obj.cp(i)  = py.CoolProp.CoolProp.PropsSI('CPMASS','P',obj.p(i)*1e5,'T',obj.T(i)+273.15,obj.fluid)/1e3;
                    end
                case {'pq','qp'}
                    for i = 1:numel(obj.p)
                        obj.s(i) = py.CoolProp.CoolProp.PropsSI('S','P',obj.p(i)*1e5,'Q',obj.q(i),obj.fluid)/1e3;
                        obj.h(i) = py.CoolProp.CoolProp.PropsSI('H','P',obj.p(i)*1e5,'Q',obj.q(i),obj.fluid)/1e3;
                        obj.T(i) = py.CoolProp.CoolProp.PropsSI('T','P',obj.p(i)*1e5,'Q',obj.q(i),obj.fluid)-273.15;
                        obj.rho(i) = py.CoolProp.CoolProp.PropsSI('D','P',obj.p(i)*1e5,'Q',obj.q(i),obj.fluid);
                        obj.cp(i)  = py.CoolProp.CoolProp.PropsSI('CPMASS','P',obj.p(i)*1e5,'Q',obj.q(i),obj.fluid)/1e3;
                    end
                case {'ph','hp'}
                    for i = 1:numel(obj.p)
                        obj.q(i) = py.CoolProp.CoolProp.PropsSI('Q','P',obj.p(i)*1e5,'H',obj.h(i)*1e3,obj.fluid);
                        obj.s(i) = py.CoolProp.CoolProp.PropsSI('S','P',obj.p(i)*1e5,'H',obj.h(i)*1e3,obj.fluid)/1e3;
                        obj.T(i) = py.CoolProp.CoolProp.PropsSI('T','P',obj.p(i)*1e5,'H',obj.h(i)*1e3,obj.fluid)-273.15;
                        obj.rho(i) = py.CoolProp.CoolProp.PropsSI('D','P',obj.p(i)*1e5,'H',obj.h(i)*1e3,obj.fluid);
                        obj.cp(i)  = py.CoolProp.CoolProp.PropsSI('CPMASS','P',obj.p(i)*1e5,'H',obj.h(i)*1e3,obj.fluid)/1e3;
                    end
                case {'Tq','qT'}
                    for i = 1:numel(obj.T)
                        obj.s(i) = py.CoolProp.CoolProp.PropsSI('S','Q',obj.q(i),'T',obj.T(i)+273.15,obj.fluid)/1e3;
                        obj.h(i) = py.CoolProp.CoolProp.PropsSI('H','Q',obj.q(i),'T',obj.T(i)+273.15,obj.fluid)/1e3;
                        obj.p(i) = py.CoolProp.CoolProp.PropsSI('P','Q',obj.q(i),'T',obj.T(i)+273.15,obj.fluid)/1e5;
                        obj.rho(i) = py.CoolProp.CoolProp.PropsSI('D','Q',obj.q(i),'T',obj.T(i)+273.15,obj.fluid);
                        obj.cp(i)  = py.CoolProp.CoolProp.PropsSI('CPMASS','Q',obj.q(i),'T',obj.T(i)+273.15,obj.fluid)/1e3;
                    end
                case 'T'
                    for i = 1:numel(obj.T)
                        obj.h(i) = interp1(obj.tab(:,1),obj.tab(:,2),obj.T(i)+273.15)/1e3;
                        obj.s(i) = interp1(obj.tab(:,1),obj.tab(:,4),obj.T(i)+273.15)/1e3;
                        obj.rho(i) = interp1(obj.tab(:,1),obj.tab(:,3),obj.T(i)+273.15);
                        obj.cp(i) = interp1(obj.tab(:,1),obj.tab(:,5),obj.T(i)+273.15)/1e3;
                    end
                otherwise
                    error('Not implemeted')
            end
        end

        % Calculate mass flow or volume flow given the other value
        function obj = calc_mdot_vdot(obj,mode)
            switch mode
                case 'vdot'
                    obj.vdot = obj.mdot / obj.rho ;
                case 'mdot'
                    obj.mdot = obj.vdot * obj.rho ;
                otherwise
                    error('Not implemented')
            end
        end
    end
end

File: Python_TEA_model/fluid_class.py
from pathlib import Path

import numpy as np

try:
    from CoolProp.CoolProp import PropsSI
except ImportError:
    PropsSI = None


class fluid_class:
    def __init__(self, fluid_name: str):
        self.fluid = fluid_name
        self.type = None
        self.handle = None
        self.fname = None
        self.tab = None

        self.p = 0.0
        self.T = 0.0
        self.h = 0.0
        self.s = 0.0
        self.q = 0.0
        self.cp = 0.0
        self.rho = 0.0
        self.mdot = 0.0
        self.vdot = 0.0

        if self.fluid == "nitrate salt":
            self.type = "tab"
            self.fname = Path("data/SolarSalt.dat")
        elif self.fluid == "chloride salt":
            self.type = "tab"
            self.fname = Path("data/ChlorideSalt.dat")
        elif self.fluid == "Therminol VP1":
            self.type = "tab"
            self.fname = Path("data/TherminolVP1.dat")
        else:
            self.type = "cp"

        if self.type == "cp":
            if PropsSI is None:
                raise ImportError("CoolProp is required for fluid_class with cp fluids")
            PropsSI("T", "P", 101325, "Q", 0, "Water")
        else:
            data_dir = Path(__file__).resolve().parent / "data"
            if not self.fname.is_absolute():
                candidate = data_dir / self.fname.name
                if candidate.exists():
                    self.fname = candidate
                elif self.fname.exists():
                    self.fname = self.fname
                else:
                    self.fname = candidate

            if not self.fname.exists():
                raise FileNotFoundError(f"Fluid data file not found: {self.fname}")

            self.tab = np.loadtxt(self.fname, comments='%', delimiter=None)

    def calc_fluid_props(self, mode: str):
        if self.type == "cp":
            if PropsSI is None:
                raise ImportError("CoolProp is required for calc_fluid_props")

            if mode in {"pT", "Tp"}:
                self.s = PropsSI("S", "P", self.p * 1e5, "T", self.T + 273.15, self.fluid) / 1e3
                self.h = PropsSI("H", "P", self.p * 1e5, "T", self.T + 273.15, self.fluid) / 1e3
                self.q = PropsSI("Q", "P", self.p * 1e5, "T", self.T + 273.15, self.fluid)
                self.rho = PropsSI("D", "P", self.p * 1e5, "T", self.T + 273.15, self.fluid)
                self.cp = PropsSI("CPMASS", "P", self.p * 1e5, "T", self.T + 273.15, self.fluid) / 1e3

            elif mode in {"pq", "qp"}:
                self.s = PropsSI("S", "P", self.p * 1e5, "Q", self.q, self.fluid) / 1e3
                self.h = PropsSI("H", "P", self.p * 1e5, "Q", self.q, self.fluid) / 1e3
                self.T = PropsSI("T", "P", self.p * 1e5, "Q", self.q, self.fluid) - 273.15
                self.rho = PropsSI("D", "P", self.p * 1e5, "Q", self.q, self.fluid)
                self.cp = PropsSI("CPMASS", "P", self.p * 1e5, "Q", self.q, self.fluid) / 1e3

            elif mode in {"ph", "hp"}:
                self.q = PropsSI("Q", "P", self.p * 1e5, "H", self.h * 1e3, self.fluid)
                self.s = PropsSI("S", "P", self.p * 1e5, "H", self.h * 1e3, self.fluid) / 1e3
                self.T = PropsSI("T", "P", self.p * 1e5, "H", self.h * 1e3, self.fluid) - 273.15
                self.rho = PropsSI("D", "P", self.p * 1e5, "H", self.h * 1e3, self.fluid)
                self.cp = PropsSI("CPMASS", "P", self.p * 1e5, "H", self.h * 1e3, self.fluid) / 1e3

            elif mode in {"Tq", "qT"}:
                self.s = PropsSI("S", "Q", self.q, "T", self.T + 273.15, self.fluid) / 1e3
                self.h = PropsSI("H", "Q", self.q, "T", self.T + 273.15, self.fluid) / 1e3
                self.p = PropsSI("P", "Q", self.q, "T", self.T + 273.15, self.fluid) / 1e5
                self.rho = PropsSI("D", "Q", self.q, "T", self.T + 273.15, self.fluid)
                self.cp = PropsSI("CPMASS", "Q", self.q, "T", self.T + 273.15, self.fluid) / 1e3

            else:
                raise ValueError(f"Mode '{mode}' not implemented for cp fluid")

        else:
            if self.tab is None:
                raise RuntimeError("Fluid tabular data is not loaded")

            if mode == "T":
                T_K = self.T + 273.15
                self.h = np.interp(T_K, self.tab[:, 0], self.tab[:, 1]) / 1e3
                self.s = np.interp(T_K, self.tab[:, 0], self.tab[:, 3]) / 1e3
                self.rho = np.interp(T_K, self.tab[:, 0], self.tab[:, 2])
                self.cp = np.interp(T_K, self.tab[:, 0], self.tab[:, 4]) / 1e3
            else:
                raise ValueError(f"Mode '{mode}' not implemented for tabular fluid")

        return self

    def calc_mdot_vdot(self, mode: str):
        if mode == "vdot":
            self.vdot = self.mdot / self.rho
        elif mode == "mdot":
            self.mdot = self.vdot * self.rho
        else:
            raise ValueError(f"Mode '{mode}' not implemented for flow conversion")

        return self
